### ISSUE TITLE
Should Range<mutable_only_range>() be true or false?

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#71]
There's currently a problem in the Range and View concepts for a type like this:

```c++
struct mutable_only_range {
  int * begin();
  int * end();
};
```

`Range<mutable_only_range>()` is `false` because we need `begin(declval<mutable_only_range>())` to compile. Since there is no `begin` that takes rvalues, this selects the `begin(const T&)` overload, which tries to call `.begin()` on a const-qualified `mutable_only_range` object. Naturally, that fails.

Since `mutable_only_range` is not a `Range`, naturally it can't be a `View`, and it really seems like this type should satisfy `View`.

We could insist that `Range<mutable_only_range>()` is false but that `Range<mutable_only_range&>()` is true. That runs into trouble in `View` because reference types are not `Semiregular`. So both `View<mutable_only_range>()` and `View<mutable_only_range &>()` will be false. That obviously wrong.

The open questions are:

1. Is my intuition that this type is both a `Range` and a `View` correct?
2. If so, how do we change things to make that so? The options I see are:
  - Add an overload of `begin` that accepts rvalues.
  - Change the definition of `IteratorType` to be like the one in range-v3: if `T` has no top-level reference (lvalue or rvalue), a `&` is added. In this way, `View<mutable_only_range>()` is true but both `View<mutable_only_range&>()` and `View<mutable_only_range&&>()` are false.

(Obvious, the same is true for `end` and `SentinelType`.)

I like the range-v3 solution since it matches my expectation, so I'm going with that one for now.

**Comment by ericniebler** [comment]
And yes, I'm aware that this is hackish and unprincipled and certain to surprise some people. We should seriously consider the other option of adding a `begin` that accepts rvalues. Arguments for and against? @CaseyCarter ?

**Comment by CaseyCarter** [comment]
Yes, this is a bit of a mess. I'd prefer perfect-forwarding `begin` and `end` over having `IteratorType<T>` not be the same as `decltype(begin(declval<T>()))` - I think that way lies madness. My gut says that defining
```
template <class R>
concept bool View() {
  return Range<R>() && Semiregular<uncvref_t<R>>() && is_view<uncvref_t<R>>::value;
}
```
will correct the View problem, but I need to give it more thought.

**Comment by ericniebler** [comment]
OK, please have a look at f25dc7a and tell me what you think.

**Comment by ericniebler** [comment]
I wonder if the rvalue overload of `begin` should behave as if the argument were a const lvalue, as in:

```c++
template <_Is<std::is_object> T>
  requires requires (T& t) { {cbegin(t)} -> auto&&; }
auto begin(T&& t) noexcept(noexcept(cbegin(t))) {
  return cbegin(t);
}
```

This is a bit safer: mutations of temporaries will be caught at compile-time. Also, the addition of this overload is no longer a breaking change.

**Comment by CaseyCarter** [comment]
I think there are two consistent approaches:

1. cv-qualifiers and value category are significant. It's possible for `const volatile Foo&&` to be a `Range` even when `Foo&` or `const Foo&&` is not. *Everything* necessarily becomes a forwarding reference, capturing a copy of a `View` requires a special wrapper that calls the appropriate `begin`/`end`. The library is fast, flexible, hard to use, and nearly impossible to extend.
2. Only cv-qualifiers are significant, and not much at that. `Range<T>` requires `begin(T&)` and `end(T&)`. `const` types may model `Range` but not `View` since they are not `Semiregular` (unless someone writes `operator = (const T&) const`, but let's not concern ourselves).`volatile` types are "don't care" as usual. The library is nearly as fast, slightly less flexible - I can't write a `Range` with only `&&`-qualified `begin` - and most importantly *much* easier to use and extend since this design hews closer to the traditional value-semantic philosophy of the standard library.

I'm strongly inclined toward the second approach - as is obvious from the biased nature of the presentation. Your understanding of how that value-semantic-ish design would affect the views and actions obviously far outstrips mine, do you think it's sane? Can we make things simpler without losing anything important?

**Comment by ericniebler** [comment]
> Your understanding of how that value-semantic-ish design would affect the views and actions obviously far outstrips mine, do you think it's sane? Can we make things simpler without losing anything important?

Good questions. We can certainly try the experiment in range-v3 and see what breaks.

**Comment by ericniebler** [comment]
> Only cv-qualifiers are significant, and not much at that. `Range<T>` requires `begin(T&)` and `end(T&)`.

The implication then is that `IteratorType<T>` and `IteratorType<T&>` are equivalent. Or did you have something else in mind?

**Comment by ericniebler** [comment]
Also, if we're disregarding top-level references for `Range<T>()`, we should be consistent and do the same for the iterator concepts, too. Right now, `Readable<I>` requires `Semiregular<I>`, so `Readable<int*&>` is false.

**Comment by CaseyCarter** [comment]
> The implication then is that IteratorType<T> and IteratorType<T&> are equivalent. Or did you have something else in mind?

Yes, `IteratorType<T>` would be `decltype(begin(declval<T&>()))`, with the same treatment for `SentinelType`. `Range<T&>` would be either equivalent to `Range<T>` for convenience or specifically excluded by requiring `Same<R, remove_reference_t<R>>` or similar. A type is or is not a range, although not every possible combination of cv-qualifier and value category is necessarily iterable.

**Comment by ericniebler** [comment]
> or specifically excluded by requiring `Same<R, remove_reference_t<R>>` or similar.

That would make constraining forwarding functions like the algorithms needlessly verbose.

**Comment by CaseyCarter** [comment]
> That would make constraining forwarding functions like the algorithms needlessly verbose.

But it would make usage of Range and View consistent. There's two sides to the issue and I'm not sure which side I'm on yet.

**Comment by ericniebler** [comment]
How would Range and View be inconsistent if we consistently ignored ref qualifiers?

**Comment by CaseyCarter** [comment]
> How would Range and View be inconsistent if we consistently ignored ref qualifiers?

I suppose because I'd like to be able to make simple statements like "a `View` is `Semiregular`," instead of "a `View` is either `Semiregular` or a possibly cv-qualified reference to another type `T` that is also a `View` and is `Semiregular`." I've never heard anyone suggest that `std::vector<int>&` should be a Container, or that `int*&&` should be an Iterator, or that `long long&` should satisfy `is_integral`. We mostly think in terms of value types. Needing to strip qualifiers from a forwarding reference to determine if it's a View or Range wouldn't impose more of a burden than needing to strip qualifiers from a forwarding reference to determine if it is_array or is_arithmetic.

**Comment by ericniebler** [comment]
This is why range-v3 uses `val` instead of `declval`. If you pass a concept a ref-qualified type, it tests exactly that type. Otherwise it treats it as `T&`. It's hack-ish, but it's often gives results you want. This needs more thought and whatever we come up with we need to apply consistently.

**Comment by CaseyCarter** [comment]
There's also a middle-ground solution: ignore ref-qualifiers only for non-value-semantic concepts like `Range` and `UniformRandomNumberGenerator` that are always passed by reference to enable convenient perfect-forwarding syntax. Value-semantic constant-time-copy concepts like `Integral` and `View` and `ForwardIterator` are typically passed by value so they have little to gain.

Also: I'm now against requiring `begin` / `end(T&&)`. I think they are unnecessary since we never actually iterate over an rvalue `Range`. By the time we get around to actually calling `begin` and `end` we're universally using an lvalue `Range` - even `for (auto&& i : function_that_returns_a_range())` binds the returned range to an rvalue reference which is an lvalue in the calls to `begin` and `end`. So I think `Range<T>()` should mean one of:
* relaxed interpretation: "If I have an expression with type and value category `T`, can I bind it to an lvalue and iterate over that lvalue?"
* strict interpretation: "Can I iterate over an lvalue of type `T`?"
In either case, `IteratorType<T>` is `decltype(begin(declval<T&>()))`.

(If we do define semantics for `begin` / `end(T&&)`, I think e.g. `begin` should return `dangling<decltype(make_move_iterator(begin(declval<T&>())))>` - but that's a separate issue. And I think not needed since `view::move` is a better solution for the same problem.)

See https://github.com/CaseyCarter/cmcstl2/tree/lvalue_range_semantics for an implementation that allows the "strict interpretation" to be selected independently for `Range` and `View`.

**Comment by ericniebler** [comment]
> Needing to strip qualifiers from a forwarding reference to determine if it's a `View` or `Range` wouldn't impose more of a burden than needing to strip qualifiers from a forwarding reference to determine if it `is_array` or `is_arithmetic`.

The difference is that now we have a concepts short-form that can't be used if we need to strip qualifiers before testing.

>  I'm now against requiring `begin` / `end(T&&)`. I think they are unnecessary since we never actually iterate over an rvalue `Range`.

If `begin`/`end` should *always* be called on lvalue objects, then `begin`/end(T&&)` should be declared `= delete`. I'm not opposed.

**Comment by CaseyCarter** [comment]
> The difference is that now we have a concepts short-form that can't be used if we need to strip qualifiers before testing.

Yes, the allure of:
```
Range{R}
DifferenceType<R> distance(R&& r) {
  return stl2::distance(begin(r), end(r));
}

SizedRange{R}
DifferenceType<R> distance(R&& r) {
  return size(r);
}
```
is hard to ignore.

> If begin/end should always be called on lvalue objects, then begin/end(T&&) should be declared = delete. I'm not opposed. EDIT: this is a breaking change.

The fact that `begin(T&&)` happens to work and return a `const` iterator is probably more a historical accident than anything else. If we had useful semantics (a la `return make_dangling(make_move_iterator(begin(T&)))`) to propose then I would say it's a reasonable breaking change. Given that we're *not* proposing "better" semantics for `begin(T&&)`/`end(T&&)` right now we should probably just leave them alone.

**Comment by CaseyCarter** [comment]
In short: I now think `Range` (and perhaps URNG) should ignore ref-qualifiers and we shouldn't waste effort on `begin`/`end(T&&)`.

**Comment by CaseyCarter** [comment]
In short: I now think Range (and perhaps URNG) should ignore ref-qualifiers and we shouldn't waste effort on begin/end(T&&).

Those semantics - upon which I think we've reached consensus - are implemented in https://github.com/CaseyCarter/cmcstl2/compare/lvalue_range_semantics. Let me know if you agree, and I'll push wording.

**Comment by ericniebler** [comment]
> Those semantics - upon which I think we've reached consensus - are implemented in https://github.com/CaseyCarter/cmcstl2/compare/lvalue_range_semantics. Let me know if you agree, and I'll push wording.

This looks good to me.

**Comment by ericniebler** [comment]
I'm going to merge that branch, FWIW.

**Comment by ericniebler** [comment]
See CaseyCarter/cmcstl2@34aada7

**Comment by ericniebler** [comment]
Fixed in 48c18ca
